### PR TITLE
Fix a bug with delegate/async SSL

### PR DESCRIPTION
**Pull request by chrisvest**
Motivation:
This bug could occasionally cause SSL handshakes to time out, because the server-side handshake would fail to resume its event loop.

Modification:
Async delegate SSL tasks now lower their NEED_TASK status after they have executed, but before they run their completion callback.
This is important because the completion callback could be querying the handshake status.
This could cause the task delegator thread and the event look to race.
If the event look queries the handshake status first, it might think that it still needs to delegate another task.
If this happens, the delegator find a null task, and then fail to resume the event loop, causing the handshake to stall.

Result:
This data race no longer causes handshake timeouts.

This is the bug that has been causing me tons of trouble on #11517

Note that existing tests already cover this scenario, just not very likely.

### CODE REVIEW

File: handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java [hunk 1]
```diff
@@ -1457,18 +1457,22 @@ public void run() {
         }
 
         @Override
-        public void run(Runnable runnable) {
+        public void run(final Runnable runnable) {
```
Comment (normanmaurer) [inline handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java:1460]: In 4.1 you will need the final tho

**Review by normanmaurer — APPROVED** [review]

File: handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java (hunk 1 above)
Comment (chrisvest) [reply handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java:1460]: This PR is for 4.1 and the `final` is added, though?

**Review by NiteshKant — APPROVED** [review]

File: handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java (hunk 1 above)
Comment (normanmaurer) [reply handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java:1460]: This is what happens when you review PRs during your vacation 🤦‍♂️... let's ship it

Comment (NiteshKant) [reply handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java:1460]: green is red after a few 🍻 norman, it is all right 😄

**Comment by johnou** [comment]
@chrisvest is this a recent regression or something that has existed from the beginning?

**Comment by chrisvest** [comment]
@johnou This regressed in **4.1.66** (specifically #11460) and you are only impacted if you use **delegate task executors**, e.g. with `SslContext.newHandler(ByteBufAllocator, String, int, Executor)`.

**Comment by johnou** [comment]
> @johnou This regressed in **4.1.66** (specifically #11460) and you are only impacted if you use **delegate task executors**, e.g. with `SslContext.newHandler(ByteBufAllocator, String, int, Executor)`, and you are using OpenSSL or BoringSSL.

Yep, we were one of the first to adopt it. Thanks for the info!

**Pull request merged**